### ISSUE TITLE
Istio Proxy is generating ~300 DNS requests per second

### ISSUE THREAD

**Issue by michaelkipper** [istio/istio#12181]
**Describe the bug**
We were paged by our DNS team that the Istio proxies are generating ~300 requests per second.

![screen shot 2019-03-01 at 9 50 25 am](https://user-images.githubusercontent.com/24932723/53645771-998b6d80-3c07-11e9-8e45-69c584a795c0.png)

**Expected behaviour**
We expected that this data would be cached, or polled slower.

**Steps to reproduce the bug**
This is a pretty basic setup.

**Version**
1.1.0-rc0

**Installation**
Custom script, using `helm template`, post-processing scripts, and [Shipit](https://github.com/Shopify/shipit-engine).

**Environment**
GKE

cc: @dwradcliffe

**Comment by michaelkipper** [comment]
cc: @duderino

**Comment by hzxuzhonghu** [comment]
I think mostly from envoy,  could add a dns cache in envoy?

**Comment by michaelkipper** [comment]
@hzxuzhonghu: Is that something Istio controls via the xds config?
Or is it something we have to enable manually?

**Comment by hzxuzhonghu** [comment]
no, istio can not control.

**Comment by rshriram** [comment]
do you know where the DNS requests are going to? is it for all services?

**Comment by rshriram** [comment]
We need to set the `dns_refresh_rate` in envoy clusters to reduce the DNS load..

**Comment by michaelkipper** [comment]
This should be resolved in https://github.com/istio/istio/pull/12272
Will wait to implement that and try it out.

cc: @dwradcliffe

**Comment by robertpanzer** [comment]
I also tested this.
I created a VirtualService and a ServiceEntry like this:
```yaml
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: dns-test
spec:
  gateways:
    - mesh
  hosts:
    - mygoogle.com
  http:
    - match:
        - gateways:
            - mesh
          port: 80
      rewrite:
        authority: google.com
      route:
        - destination:
            host: google.com
            port:
              number: 80
          weight: 100

---

apiVersion: networking.istio.io/v1alpha3
kind: ServiceEntry
metadata:
  name: dns-test
spec:
  hosts:
    - google.com
  location: MESH_EXTERNAL
  ports:
    - number: 80
      name: http
      protocol: HTTP
  resolution: DNS
  endpoints:
    - address: google.com
      ports:
        http: 80
```
(Looks weird, but it just routes the host name mygoogle.com to google.com)

Then I started taking a tcpdump from the proxy with this command: 
```
sudo tcpdump -lvi any "udp port 53" |grep google.com
```
I see that Envoy seems to do the DNS lookups every 5 seconds due to this PR not being installed yet.
But what I found quite strange is that it always tries to do the lookups for the internal names:
```
    testcurl-7c7454f99c-glqdg.60141 > kube-dns.kube-system.svc.cluster.local.53: 20535+ A? google.com.sr-test.svc.cluster.local. (54)
    testcurl-7c7454f99c-glqdg.60141 > kube-dns.kube-system.svc.cluster.local.53: 43839+ A? google.com.svc.cluster.local. (46)
    testcurl-7c7454f99c-glqdg.60141 > kube-dns.kube-system.svc.cluster.local.53: 15298+ A? google.com.cluster.local. (42)
    testcurl-7c7454f99c-glqdg.60141 > kube-dns.kube-system.svc.cluster.local.53: 27378+ A? google.com.foo.com. (40)
    testcurl-7c7454f99c-glqdg.60141 > kube-dns.kube-system.svc.cluster.local.53: 38695+ A? google.com. (28)
    kube-dns.kube-system.svc.cluster.local.53 > testcurl-7c7454f99c-glqdg.60141: 38695 6/0/0 google.com. A 108.177.122.113, google.com. A 108.177.122.100, google.com. A 108.177.122.138, google.com. A 108.177.122.101, google.com. A 108.177.122.102, google.com. A 108.177.122.139 (184)
```
I configured the ServiceEntry to be MESH_EXTERNAL, so there should be no point looking for google.com.svc.cluster.local.
I guess this is just coming from the configuration in `/etc/resolv.conf`.

Envoy would allow setting a different DNS resolver, but that would mean that there is some kind DNS server in Istio.

Looking at the order of resolves it seems also to be preferable to reference internal services with the short names, e.g. `<servicename>` or `<servicename>.<namespace>` instead of the full name

**Comment by rshriram** [comment]
@robertpanzer thats because of the default DNS search namespaces in the pod. In kubernetes, it defaults to searching in local namespace, then all namespaces, and then to external DNS.

**Comment by duderino** [comment]
https://github.com/istio/istio/pull/12272 will be in 1.1rc3 which should be out in a few hours

**Comment by duderino** [comment]
@mandarjog this would be a good thing to try in one of the perf clusters

**Comment by mikesimons** [comment]
We just discovered this on Istio 1.1 when we pushed 12 ServiceEntries w/ 50 ports each to our dev cluster which has 12 namespaces & ~13 pods each.

coredns started getting OOMKilled & entering CrashLoopBackoff because the envoy sidecars were flooding them with requests. Scaling out the replicas and increasing the memory limit on coredns allowed us to troubleshoot. We've worked around it by lowering the dnsRefreshInterval to 30s and reducing the number of ports to the strictly necessary ones but the cluster was completely broken while we figured it out.

We're migrating to an EKS cluster from a port based routing setup (hence the silly number of ports). Even though there are only 12 unique hostnames among the ServiceEntries envoy sends far fewer requests when we reduce the port count leading us to believe that it may be duplicating lookups.

Could doing the DNS lookups at pilot and pushing them out to envoy as STATIC be an option?
